File: backend/api/comparisons.py
"""
Resume comparison endpoints for multi-resume analysis and ranking.

This module provides endpoints for creating, retrieving, and managing
multi-resume comparison views with ranking, filtering, and sorting capabilities.
"""
import logging
from typing import List, Optional

from fastapi import APIRouter, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

router = APIRouter()


class ComparisonCreate(BaseModel):
    """Request model for creating a comparison view."""

    vacancy_id: str = Field(..., description="ID of the job vacancy to compare against")
    resume_ids: List[str] = Field(..., description="List of resume IDs to compare (2-5 resumes)", min_length=2, max_length=5)
    name: Optional[str] = Field(None, description="Optional name for the comparison view")
    filters: Optional[dict] = Field(None, description="Filter settings (match range, sort field, etc.)")
    created_by: Optional[str] = Field(None, description="User identifier who created the comparison")
    shared_with: Optional[List[str]] = Field(None, description="List of user IDs/emails to share with")


class ComparisonUpdate(BaseModel):
    """Request model for updating a comparison view."""

    name: Optional[str] = Field(None, description="Updated name for the comparison view")
    filters: Optional[dict] = Field(None, description="Updated filter settings")
    shared_with: Optional[List[str]] = Field(None, description="Updated list of users to share with")


class ComparisonResponse(BaseModel):
    """Response model for a comparison view."""

    id: str = Field(..., description="Unique identifier for the comparison")
    vacancy_id: str = Field(..., description="ID of the job vacancy")
    resume_ids: List[str] = Field(..., description="List of resume IDs being compared")
    name: Optional[str] = Field(None, description="Name of the comparison view")
    filters: Optional[dict] = Field(None, description="Filter settings")
    created_by: Optional[str] = Field(None, description="User who created the comparison")
    shared_with: Optional[List[str]] = Field(None, description="List of users shared with")
    comparison_results: Optional[List[dict]] = Field(None, description="Match results for each resume")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")


class ComparisonListResponse(BaseModel):
    """Response model for listing comparison views."""

    comparisons: List[ComparisonResponse] = Field(..., description="List of comparison views")
    total_count: int = Field(..., description="Total number of comparison views")


@router.post(
    "/",
    response_model=ComparisonResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["Comparisons"],
)
async def create_comparison(request: ComparisonCreate) -> JSONResponse:
    """
    Create a new resume comparison view.

    This endpoint creates a new comparison view for analyzing multiple resumes
    side-by-side against a job vacancy. The comparison can be saved with filters,
    shared with team members, and retrieved later.

    Args:
        request: Create request with vacancy_id, resume_ids, and optional settings

    Returns:
        JSON response with created comparison view details

    Raises:
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {
        ...     "vacancy_id": "vacancy-123",
        ...     "resume_ids": ["resume1", "resume2", "resume3"],
        ...     "name": "Senior Developer Candidates"
        ... }
        >>> response = requests.post("http://localhost:8000/api/comparisons/", json=data)
        >>> response.json()
        {
            "id": "comp-123",
            "vacancy_id": "vacancy-123",
            "resume_ids": ["resume1", "resume2", "resume3"],
            "name": "Senior Developer Candidates",
            "filters": None,
            "created_by": None,
            "shared_with": None,
            "comparison_results": None,
            "created_at": "2024-01-25T00:00:00Z",
            "updated_at": "2024-01-25T00:00:00Z"
        }
    """
    try:
        logger.info(
            f"Creating comparison for vacancy_id: {request.vacancy_id} "
            f"with {len(request.resume_ids)} resumes"
        )

        # Validate resume_ids count
        if len(request.resume_ids) < 2:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="At least 2 resumes must be provided for comparison",
            )
        if len(request.resume_ids) > 5:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Maximum 5 resumes can be compared at once",
            )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        comparison_response = {
            "id": "placeholder-comparison-id",
            "vacancy_id": request.vacancy_id,
            "resume_ids": request.resume_ids,
            "name": request.name,
            "filters": request.filters,
            "created_by": request.created_by,
            "shared_with": request.shared_with,
            "comparison_results": None,
            "created_at": "2024-01-25T00:00:00Z",
            "updated_at": "2024-01-25T00:00:00Z",
        }

        logger.info(
            f"Created comparison for vacancy {request.vacancy_id} "
            f"with {len(request.resume_ids)} resumes"
        )

        return JSONResponse(
            status_code=status.HTTP_201_CREATED,
            content=comparison_response,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating comparison: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create comparison: {str(e)}",
        ) from e


@router.get("/", tags=["Comparisons"])
async def list_comparisons(
    vacancy_id: Optional[str] = Query(None, description="Filter by vacancy ID"),
    created_by: Optional[str] = Query(None, description="Filter by creator user ID"),
    limit: int = Query(50, description="Maximum number of comparisons to return", ge=1, le=100),
    offset: int = Query(0, description="Number of comparisons to skip", ge=0),
) -> JSONResponse:
    """
    List resume comparison views with optional filters.

    Args:
        vacancy_id: Optional vacancy ID filter
        created_by: Optional creator user ID filter
        limit: Maximum number of results to return (default: 50, max: 100)
        offset: Number of results to skip (default: 0)

    Returns:
        JSON response with list of comparison views

    Raises:
        HTTPException(500): If database query fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/comparisons/?vacancy_id=vac-123")
        >>> response.json()
    """
    try:
        logger.info(
            f"Listing comparisons with filters - vacancy_id: {vacancy_id}, "
            f"created_by: {created_by}, limit: {limit}, offset: {offset}"
        )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        response_data = {
            "comparisons": [],
            "total_count": 0,
        }

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error listing comparisons: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list comparisons: {str(e)}",
        ) from e


@router.get("/{comparison_id}", tags=["Comparisons"])
async def get_comparison(comparison_id: str) -> JSONResponse:
    """
    Get a specific comparison view by ID.

    Args:
        comparison_id: Unique identifier of the comparison view

    Returns:
        JSON response with comparison view details

    Raises:
        HTTPException(404): If comparison view is not found
        HTTPException(500): If database query fails

    Examples:
        >>> import requests
        >>> response = requests.get(
        ...     "http://localhost:8000/api/comparisons/123e4567-e89b-12d3-a456-426614174000"
        ... )
        >>> response.json()
    """
    try:
        logger.info(f"Getting comparison: {comparison_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "id": comparison_id,
                "vacancy_id": "vacancy-123",
                "resume_ids": ["resume1", "resume2"],
                "name": "Sample Comparison",
                "filters": None,
                "created_by": "user-123",
                "shared_with": None,
                "comparison_results": None,
                "created_at": "2024-01-25T00:00:00Z",
                "updated_at": "2024-01-25T00:00:00Z",
            },
        )

    except Exception as e:
        logger.error(f"Error getting comparison: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get comparison: {str(e)}",
        ) from e


@router.put("/{comparison_id}", tags=["Comparisons"])
async def update_comparison(
    comparison_id: str, request: ComparisonUpdate
) -> JSONResponse:
    """
    Update a comparison view.

    Args:
        comparison_id: Unique identifier of the comparison view
        request: Update request with fields to modify

    Returns:
        JSON response with updated comparison view

    Raises:
        HTTPException(404): If comparison view is not found
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {"name": "Updated Comparison Name"}
        >>> response = requests.put(
        ...     "http://localhost:8000/api/comparisons/123",
        ...     json=data
        ... )
        >>> response.json()
    """
    try:
        logger.info(f"Updating comparison: {comparison_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "id": comparison_id,
                "vacancy_id": "vacancy-123",
                "resume_ids": ["resume1", "resume2"],
                "name": request.name if request.name is not None else "Sample Comparison",
                "filters": request.filters,
                "created_by": "user-123",
                "shared_with": request.shared_with,
                "comparison_results": None,
                "created_at": "2024-01-25T00:00:00Z",
                "updated_at": "2024-01-25T00:00:00Z",
            },
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating comparison: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update comparison: {str(e)}",
        ) from e


@router.delete("/{comparison_id}", tags=["Comparisons"])
async def delete_comparison(comparison_id: str) -> JSONResponse:
    """
    Delete a comparison view.

    Args:
        comparison_id: Unique identifier of the comparison view

    Returns:
        JSON response confirming deletion

    Raises:
        HTTPException(404): If comparison view is not found
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> response = requests.delete("http://localhost:8000/api/comparisons/123")
        >>> response.json()
        {"message": "Comparison deleted successfully"}
    """
    try:
        logger.info(f"Deleting comparison: {comparison_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={"message": f"Comparison {comparison_id} deleted successfully"},
        )

    except Exception as e:
        logger.error(f"Error deleting comparison: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete comparison: {str(e)}",
        ) from e

File: frontend/src/api/comparisons.ts
/**
 * Comparisons API Client
 *
 * Этот модуль предоставляет клиент для работы со сравнениями резюме.
 * Поддерживает полный цикл управления сравнениями: создание, просмотр списка,
 * просмотр деталей, обновление, удаление и множественное сравнение резюме с вакансией.
 *
 * @example
 * ```ts
 * import { comparisonsClient, ComparisonsClient } from '@/api/comparisons';
 *
 * // Создание нового сравнения
 * const comparison = await comparisonsClient.createComparison({
 *   vacancy_id: 'vacancy-123',
 *   resume_ids: ['resume-1', 'resume-2', 'resume-3'],
 *   name: 'Top Candidates for Senior Developer',
 * });
 *
 * // Получение списка сравнений с фильтрацией
 * const list = await comparisonsClient.listComparisons('vacancy-123');
 *
 * // Сравнение множественных резюме
 * const matrix = await comparisonsClient.compareMultipleResumes({
 *   vacancy_id: 'vacancy-123',
 *   resume_ids: ['resume-1', 'resume-2'],
 * });
 *
 * // Обновление сравнения
 * const updated = await comparisonsClient.updateComparison('comparison-123', {
 *   name: 'Updated Name',
 * });
 *
 * // Удаление сравнения
 * await comparisonsClient.deleteComparison('comparison-123');
 * ```
 */

import axios, { AxiosInstance, AxiosError, AxiosResponse } from 'axios';
import type {
  ComparisonCreate,
  ComparisonUpdate,
  ComparisonResponse,
  ComparisonListResponse,
  CompareMultipleRequest,
  ComparisonMatrixData,
  ApiError,
} from '@/types/api';

/**
 * Переэкспорт типов для удобства использования
 */
export type {
  ComparisonCreate,
  ComparisonUpdate,
  ComparisonResponse,
  ComparisonListResponse,
  CompareMultipleRequest,
  ComparisonMatrixData,
};

/**
 * Конфигурация по умолчанию для клиента сравнений
 */
const DEFAULT_CONFIG = {
  baseURL: import.meta.env.VITE_API_GATEWAY_URL ?? 'http://localhost:8888',
  timeout: 30000,
  headers: {
    'Content-Type': 'application/json',
  },
};

/**
 * Класс клиента API для работы со сравнениями резюме
 *
 * Предоставляет методы для создания, просмотра, обновления и удаления сравнений
 * с proper обработкой ошибок и типобезопасностью.
 */
export class ComparisonsClient {
  private client: AxiosInstance;

  /**
   * Создание нового экземпляра клиента сравнений
   *
   * @param config - Опциональные переопределения конфигурации
   */
  constructor(config: Partial<typeof DEFAULT_CONFIG> = {}) {
    const finalConfig = { ...DEFAULT_CONFIG, ...config };

    this.client = axios.create(finalConfig);

    // Интерцептор ответов для обработки ошибок
    this.client.interceptors.response.use(
      (response) => response,
      (error) => Promise.reject(this.transformError(error))
    );
  }

  /**
   * Преобразование ошибки Axios в стандартизированную ошибку API
   *
   * @param error - Ошибка Axios
   * @returns Преобразованная ошибка API
   */
  private transformError(error: unknown): ApiError {
    const axiosError = error as AxiosError<{ detail?: string }>;

    // Ошибка сети (нет ответа)
    if (!axiosError.response) {
      if (axiosError.code === 'ECONNABORTED') {
        return {
          detail: 'Таймаут запроса. Проверьте соединение и попробуйте снова.',
          status: 408,
        };
      }
      return {
        detail: 'Ошибка сети. Проверьте соединение и попробуйте снова.',
        status: 0,
      };
    }

    // Сервер вернул ошибку
    const status = axiosError.response.status;
    const data = axiosError.response.data;

    // Используем сообщение об ошибке от сервера, если доступно
    if (data?.detail) {
      return { detail: data.detail, status };
    }

    // Сообщения об ошибках по умолчанию для разных кодов статуса
    const defaultMessages: Record<number, string> = {
      400: 'Неверный запрос. Проверьте введенные данные.',
      401: 'Не авторизован. Войдите в систему.',
      403: 'Доступ запрещен. У вас нет прав для выполнения этого действия.',
      404: 'Сравнение не найдено.',
      422: 'Ошибка валидации. Проверьте формат данных.',
      429: 'Слишком много запросов. Попробуйте позже.',
      500: 'Ошибка сервера. Попробуйте позже.',
      502: 'Ошибка шлюза. Попробуйте позже.',
      503: 'Сервис недоступен. Попробуйте позже.',
    };

    return {
      detail: data?.detail || defaultMessages[status] || 'Произошла непредвиденная ошибка.',
      status,
    };
  }

  /**
   * Создание нового сравнения резюме
   *
   * @param request - Запрос на создание сравнения
   * @returns Созданное сравнение
   * @throws ApiError если создание не удалось
   *
   * @example
   * ```ts
   * const comparison = await comparisonsClient.createComparison({
   *   vacancy_id: 'vacancy-123',
   *   resume_ids: ['resume-1', 'resume-2', 'resume-3'],
   *   name: 'Top Candidates for Senior Developer',
   *   filters: { min_experience: 5 },
   * });
   * ```
   */
  async createComparison(request: ComparisonCreate): Promise<ComparisonResponse> {
    try {
      const response: AxiosResponse<ComparisonResponse> = await this.client.post(
        '/api/comparisons/',
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Получение списка сравнений с опциональными фильтрами и сортировкой
   *
   * @param vacancyId - Опциональный фильтр по ID вакансии
   * @param createdBy - Опциональный фильтр по ID создателя
   * @param minMatchPercentage - Опциональный минимальный процент совпадения (0-100)
   * @param maxMatchPercentage - Опциональный максимальный процент совпадения (0-100)
   * @param sortBy - Поле для сортировки: created_at, match_percentage, name, updated_at
   * @param order - Порядок сортировки: asc или desc
   * @param limit - Максимальное количество результатов
   * @param offset - Смещение для пагинации
   * @returns Список сравнений с метаданными
   * @throws ApiError если получение списка не удалось
   *
   * @example
   * ```ts
   * // Получение всех сравнений для вакансии
   * const list = await comparisonsClient.listComparisons('vacancy-123');
   *
   * // Фильтрация по проценту совпадения с сортировкой
   * const filtered = await comparisonsClient.listComparisons(
   *   'vacancy-123',
   *   undefined,
   *   80,
   *   100,
   *   'match_percentage',
   *   'desc'
   * );
   *
   * // Пагинация
   * const paginated = await comparisonsClient.listComparisons(
   *   undefined,
   *   undefined,
   *   undefined,
   *   undefined,
   *   undefined,
   *   undefined,
   *   10,
   *   20
   * );
   * ```
   */
  async listComparisons(
    vacancyId?: string,
    createdBy?: string,
    minMatchPercentage?: number,
    maxMatchPercentage?: number,
    sortBy?: string,
    order?: string,
    limit?: number,
    offset?: number
  ): Promise<ComparisonListResponse> {
    try {
      const params: Record<string, string | number> = {};
      if (vacancyId) params.vacancy_id = vacancyId;
      if (createdBy) params.created_by = createdBy;
      if (minMatchPercentage !== undefined) params.min_match_percentage = minMatchPercentage;
      if (maxMatchPercentage !== undefined) params.max_match_percentage = maxMatchPercentage;
      if (sortBy) params.sort_by = sortBy;
      if (order) params.order = order;
      if (limit !== undefined) params.limit = limit;
      if (offset !== undefined) params.offset = offset;

      const response: AxiosResponse<ComparisonListResponse> = await this.client.get(
        '/api/comparisons/',
        { params }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Получение сравнения по ID
   *
   * @param id - ID сравнения
   * @returns Детали сравнения
   * @throws ApiError если получение не удалось
   *
   * @example
   * ```ts
   * const comparison = await comparisonsClient.getComparison('comparison-123');
   * console.log(`Сравнение для вакансии: ${comparison.vacancy_id}`);
   * console.log(`Количество резюме: ${comparison.resume_ids.length}`);
   * ```
   */
  async getComparison(id: string): Promise<ComparisonResponse> {
    try {
      const response: AxiosResponse<ComparisonResponse> = await this.client.get(
        `/api/comparisons/${id}`
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Обновление сравнения
   *
   * @param id - ID сравнения
   * @param request - Запрос на обновление с полями для изменения
   * @returns Обновленное сравнение
   * @throws ApiError если обновление не удалось
   *
   * @example
   * ```ts
   * const updated = await comparisonsClient.updateComparison('comparison-123', {
   *   name: 'Updated Comparison Name',
   *   filters: { min_experience: 7 },
   *   shared_with: ['user-1', 'user-2'],
   * });
   * ```
   */
  async updateComparison(
    id: string,
    request: ComparisonUpdate
  ): Promise<ComparisonResponse> {
    try {
      const response: AxiosResponse<ComparisonResponse> = await this.client.put(
        `/api/comparisons/${id}`,
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Удаление сравнения
   *
   * @param id - ID сравнения
   * @throws ApiError если удаление не удалось
   *
   * @example
   * ```ts
   * await comparisonsClient.deleteComparison('comparison-123');
   * ```
   */
  async deleteComparison(id: string): Promise<void> {
    try {
      await this.client.delete(`/api/comparisons/${id}`);
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Сравнение множественных резюме с вакансией
   *
   * Этот метод выполняет интеллектуальное сопоставление навыков каждого резюме
   * с требованиями вакансии, обрабатывая синонимы (например, PostgreSQL ≈ SQL)
   * и предоставляя агрегированные результаты с ранжированием по проценту совпадения.
   *
   * @param request - Запрос на сравнение с vacancy_id и resume_ids
   * @returns Данные матрицы сравнения с ранжированными результатами
   * @throws ApiError если сравнение не удалось
   *
   * @example
   * ```ts
   * const matrix = await comparisonsClient.compareMultipleResumes({
   *   vacancy_id: 'vacancy-123',
   *   resume_ids: ['resume-1', 'resume-2', 'resume-3'],
   * });
   *
   * console.log(`Вакансия: ${matrix.vacancy_title}`);
   * console.log(`Обработано резюме: ${matrix.total_resumes}`);
   * console.log(`Время обработки: ${matrix.processing_time_ms}ms`);
   *
   * matrix.comparison_results.forEach((result) => {
   *   console.log(`Ранг ${result.rank}: ${result.resume_id} - ${result.match_percentage}%`);
   * });
   * ```
   */
  async compareMultipleResumes(request: CompareMultipleRequest): Promise<ComparisonMatrixData> {
    try {
      const response: AxiosResponse<ComparisonMatrixData> = await this.client.post(
        '/api/comparisons/compare-multiple',
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }
}

/**
 * Синглтон экземпляр клиента сравнений для удобного использования
 *
 * @example
 * ```ts
 * import { comparisonsClient } from '@/api/comparisons';
 *
 * const comparison = await comparisonsClient.createComparison({
 *   vacancy_id: 'vacancy-123',
 *   resume_ids: ['resume-1', 'resume-2'],
 * });
 * ```
 */
export const comparisonsClient = new ComparisonsClient();
